src/main/java/com/squareup/javapoet/AnnotationSpec.java

```diff
@@ -221,7 +222,9 @@ public Builder addMember(String name, CodeBlock codeBlock) {
      * depending on the given {@code value} object. Falls back to {@code "$L"} literal format if
      * the class of the given {@code value} object is not supported.
      */
-    private Builder addValue(String memberName, Object value) {
+    public Builder addMemberForValue(String memberName, Object value) {
```

Comment:
Maybe the problem is rooted [here](https://github.com/square/javapoet/blob/master/src/main/java/com/squareup/javapoet/AnnotationSpec.java#L199) in the declaration of the member map : `Map<String, List<CodeBlock>> members`
If it was a key-pair consisting of the member type and member name (`ParameterSpec`?)... we could also check for valid types and look up the correct format string like `$S`, `$T` and so on.